File: javascript/lesson 3/separateChainingHashtable.js
const INITIAL_CAPACITY = 4;

class Node {
    constructor(key, value, next = null) {
        this.key = key;
        this.value = value;
        this.next = next;
    }
}

class SeparateChainingHashTable {
    constructor(capacity = INITIAL_CAPACITY) {
        this.count = 0;
        this.capacity = capacity;
        this.buckets = new Array(capacity);
    }

    put(key, value) {
        if (this.count >= 10 * this.capacity) this._resize(2 * this.capacity);

        const i = this._hash(key);
        for (let node = this.buckets[i]; node != null; node = node.next) {
            if (key === node.key) {
                node.value = value;
                return;
            }
        }
        this.buckets[i] = new Node(key, value, this.buckets[i]);
        this.count++;
    }

    get(key) {
        const i = this._hash(key);
        for (let node = this.buckets[i]; node != null; node = node.next) {
            if (key === node.key) return node.value;
        }
        return null;
    }

    remove(key) {
        var i = this._hash(key);
        for (let prev = null, node = this.buckets[i]; node != null; prev = node, node = node.next) {
            if (key === node.key) {
                if (!prev) {
                    this.buckets[i] = node.next;
                } else {
                    prev.next = node.next;
                }
                this.count--;
                break;
            }
        }
        if (this.capacity > INITIAL_CAPACITY && this.count <= 2 * this.capacity) this._resize(this.capacity / 2);
    }

    _resize(size) {
        const temp = new SeparateChainingHashTable(size);
        for (let i = 0; i < this.capacity; i++) {
            for (let node = this.buckets[i]; node != null; node = node.next) {
                temp.put(node.key, node.value);
            }
        }
        this.buckets = temp.buckets;
        this.capacity = size;
    }

    _hash(key) {
        return key % this.capacity;
    }
}

File: python/lesson 3/separate_chaining_hash_table.py
class Node:
    def __init__(self, key, value, nxt):
        self.key = key
        self.value = value
        self.next = nxt

class SeparateChainingHashTable:
    INITIAL_CAPACITY = 4
    
    def __init__(self, capactiy = INITIAL_CAPACITY):
        self.count = 0
        self.capacity = capactiy
        self.buckets =  [None for _ in range(self.capacity)]

    def put(self, key, value):
        if self.count >= 10 * self.capacity: 
            self.resize(2 * self.capacity)
        i = self.__hash(key)
        node = self.buckets[i]
        while node:
            if key == node.key:
                node.value = value                
                return
            node = node.next
        self.buckets[i] = Node(key, value, self.buckets[i])
        self.count += 1

    def get(self, key):
        i = self.__hash(key)
        node = self.buckets[i]
        while node:
            if node.key == key: return node.value
            node = node.next
        return None

    def remove(self, key):
        i = self.__hash(key)
        node = self.buckets[i]
        prev = None
        while node:
            if key == node.key:
                if not prev:
                    self.buckets[i] = node.next
                else:
                    prev.next = node.next                
                self.count -= 1
                break            
            prev = node
            node = node.next
        if self.capacity > self.INITIAL_CAPACITY and self.count <= 2 * self.capacity:
            self.resize(self.capacity / 2) 

    def resize(self, size: int) -> None:
        temp = SeparateChainingHashTable(size)
        for i in range(self.capacity):
            node = self.buckets[i]
            while node:
                temp.put(node.key, node.value)
                node = node.next
        self.buckets = temp.buckets
        self.capacity = size

    def __hash(self, key):
        return key % self.capacity
